crates/runtime/src/flight/auth.rs

```diff
@@ -0,0 +1,225 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use base64::{prelude::BASE64_STANDARD, Engine};
+use pin_project::pin_project;
+use runtime_auth::{AuthVerdict, FlightBasicAuth};
+use tonic::{
+    metadata::{MetadataMap, GRPC_CONTENT_TYPE},
+    Status,
+};
+use tower::{Layer, Service};
+
+/// The handshake request for Flight basic auth is a base64 encoded token that is calculated like:
+///
+/// ```rust,no_run
+/// let val = BASE64_STANDARD.encode(format!("{username}:{password}"));
```

Comment:
Are we enabling Basic Auth anywhere in the project?

Reply:
Basic Auth is a fairly standard way for clients to connect to FlightSQL endpoints. Its how our SDKs work for API key auth, and how the FlightSQL JDBC driver works https://arrow.apache.org/docs/java/flight_sql_jdbc_driver.html